### PR TITLE
add trap: regexp matching and filter

### PR DESCRIPTION
**Pull request by garnier-quentin**
<h1> Pull Request Template </h1>

<h2> Description </h2>

Update PR:
https://github.com/centreon/centreon/pull/5742
https://github.com/centreon/centreon/pull/5983

I wait the merge to update documentation.

<h2> Type of change </h2>

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

<h2> Target serie </h2>

- [ ] 2.8.x
- [ ] 18.10.x
- [x] 19.04.x (master)

<h2> How this pull request can be tested ? </h2>

Need the last version. You can create a trap with matching mode regexp (oid value: .*)
And you send a trap to test if it's matching (for example):
snmptrap -c public -v 2c 127.0.0.1 '' .1.3.6.1.4.1.998 .1 s ''

<h2> Checklist </h2>

<h5> Community contributors & Centreon team </h5>

- [x] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

<h5> Centreon team only </h5>

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests covers 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: lib/perl/centreon/script/centreontrapd.pm [hunk 1]
```diff
@@ -625,6 +624,14 @@ sub manage_exec {
             $self->{logger}->writeLogInfo("Skipping trap '" . $self->{current_trap_id} . "' for host ID '" . $self->{current_host_id} . "': time interval");
             return 1;
         }
+        
+        # Host/Service type
+        if ($self->{trap_data}->{ref_oids}->{ $self->{current_trap_id} }->{traps_exec_interval_type} == 3 &&
+            defined($self->{last_time_exec}{host}->{$self->{current_host_id} . ";" . $self->{current_service_id} . ";" . $self->{current_oid}}) &&
+            $self->{trap_data}->{trap_date_time_epoch} < ($self->{last_time_exec}{host}->{$self->{current_host_id} . ";" . $self->{current_service_id} . ";" . $self->{current_oid}} + $self->{trap_data}->{ref_oids}->{ $self->{current_trap_id} }->{traps_exec_interval})) {
+            $self->{logger}->writeLogInfo("Skipping trap '" . $self->{current_trap_id} . "' for host ID '" . $self->{current_host_id} . "' and service ID '".$self->{current_service_id}."': time interval");
```
Comment (sc979) [inline lib/perl/centreon/script/centreontrapd.pm]:
```suggestion
            $self->{logger}->writeLogInfo("Skipping trap '" . $self->{current_trap_id} . "' for host ID '" . $self->{current_host_id} . "' and service ID '" . $self->{current_service_id}."': time interval");
```

File: lib/perl/centreon/script/centreontrapd.pm [hunk 2]
```diff
@@ -664,6 +671,7 @@ sub manage_exec {
     $self->{running_processes}->{$current_pid} = 1;
     $self->{last_time_exec}{oid}->{$self->{current_oid}} = $self->{trap_data}->{trap_date_time_epoch};
     $self->{last_time_exec}{host}->{$self->{current_host_id} . ";" . $self->{current_oid}} = $self->{trap_data}->{trap_date_time_epoch};
+    $self->{last_time_exec}{host}->{$self->{current_host_id} . ";" . $self->{current_service_id} . ";" . $self->{current_oid}} = $self->{trap_data}->{trap_date_time_epoch};
```
Comment (sc979) [inline lib/perl/centreon/script/centreontrapd.pm]: line too long

File: www/include/configuration/configObject/traps/formTraps.php [hunk 3]
```diff
@@ -63,36 +63,38 @@ function myReplace()
...
-    ));
-    $cdata->addJsData('clone-count-matchingrules', count($mrulesArray));
 }
 
+/*
+* Preset values of preexec commands
+*/
+$cdata->addJsData('clone-values-preexec', htmlspecialchars(
+     json_encode($preexecArray),
+    ENT_QUOTES
```
Comment (sc979) [inline www/include/configuration/configObject/traps/formTraps.php]: wrong indentation

File: www/include/configuration/configObject/traps/formTraps.php [hunk 4]
```diff
@@ -338,16 +345,26 @@ function myReplace()
     _("By OID and Host"),
     '2'
 );
+$excecution_type[] = $form->createElement(
+    'radio',
+    'traps_exec_interval_type',
+    null,
+    _("By OID, Host and Service"),
```
Comment (sc979) [inline www/include/configuration/configObject/traps/formTraps.php]: Missing translation

File: www/include/configuration/configObject/traps/help.php [hunk 5]
```diff
@@ -39,6 +39,10 @@
     "help",
     "Enter the trap name as specified in the MIB file and send by the SNMP master agent."
 );
+$help["mode"] = dgettext(
+    "help",
+    "Choose the matching mode."
```
Comment (sc979) [inline www/include/configuration/configObject/traps/help.php]: missing translation

File: www/install/createTables.sql [hunk 6]
```diff
@@ -2050,7 +2051,7 @@ CREATE TABLE `traps` (
   `traps_advanced_treatment_default` enum('0','1', '2') DEFAULT '0',
   `traps_timeout` int(11) DEFAULT NULL,
   `traps_exec_interval` int(11) DEFAULT NULL,
-  `traps_exec_interval_type` enum('0','1', '2') DEFAULT '0',
+  `traps_exec_interval_type` enum('0','1','2','3') DEFAULT '0',
```
Comment (sc979) [inline www/install/createTables.sql]: missing documentation in your PR. At what correspond the type 3 ?

**Comment by sc979** [comment]
missing online documentation and updated screenshot

File: www/install/createTables.sql (hunk 6 above)
Comment (garnier-quentin) [reply www/install/createTables.sql]: I'm waiting the merge before doing the doc

Comment (sc979) [reply www/install/createTables.sql]: Why wait the merge and make another PR when this is linked and you may forgot ?

**Comment by garnier-quentin** [comment]
Done!

**Comment by lpinsivy** [comment]
@garnier-quentinit will be better to put the documentation in this PR?

**Comment by garnier-quentin** [comment]
I'm missing the clapi also.

**Comment by garnier-quentin** [comment]
done

**Pull request closed without merging**